vere/serf.c

```diff
@@ -394,6 +394,8 @@ _serf_sure(u3_noun ovo, u3_noun vir, u3_noun cor)
     }
   }
 
+  //  XX this runs on replay too
```

Comment:
We typically want to avoid handling effects during event replay. I don't think this is handled quite right on cc-release yet, but it's conceptually straight forward: all the effects are sent over the pipe from the serf back to the lord, and once the time is right the effects are "released" in pier.c. The `|mass` effect is an exception to this pattern, as it is (and must be) applied directly in the serf. If we want to avoid actually garbage-collecting the entire Loom on every `|mass` during event replay, we'll need to some special handling here.

Reply:
Ah! I see. Usually the serf just ignores effects and passes them to the king.

Are you sure we shouldn't just do that here as well, to keep things consistent, and then handle the complexity from the King? Or is there too much information in the effect that only makes sense in the serf?

I'm imagining that serf just sends the |mass effect to the king, and then the king explicitly sends a requests to the serf to weigh itself. My understanding is that this doesn't work because we aren't weighing the serf, we're weighing some specific data on the serf's loom.

Reply:
The implementation of `|mass` is deeply tied into garbage collection, which is stateful and deeply coupled to the details of the Loom at that point in time.

The king could send a `%mass` or `%wegh` command to the serf, which could then statefully garbage-collect, but it would be missing the specificity contained in the effect itself. Unless the serf hung on to the effect, but that's effectively modularity theater.

One could other approaches to memory measurement (for instance, `(met 3 (jam effect))`), but this is the one we have.

I think we'll eventually want a specific replay mode for the serf (an option in the `%boot` command, perhaps) to avoid the overhead of jamming and sending a bunch of effects that we will then just ignore. In that mode, I think we would just skip `|mass`.

Reply:
Another option would be to just ignore `|mass` effects that come from the king? Would that work?

Like, you said the effect only really makes sense immediately when it's emitted, and it has no effect on the state, right?